CRM/MembershipExtras/Hook/BuildForm/MembershipPaymentPlan.php

```diff
@@ -0,0 +1,72 @@
+<?php
+
+/**
+ * Implements form changes needed to be done to add payment plan as an option to
+ * pay for a membership.
+ */
+abstract class CRM_MembershipExtras_Hook_BuildForm_MembershipPaymentPlan {
+  /**
+   * @var string
+   *   Path where template with new fields is stored.
+   */
+  protected $templatePath = '';
+
+  /**
+   * @var \CRM_Member_Form_Membership
+   *   Form object that is being altered.
+   */
+  protected $form;
+
+  /**
+   * CRM_MembershipExtras_Hook_BuildForm_Membership constructor.
+   *
+   * @param \CRM_Member_Form $form
+   */
+  public function __construct(CRM_Member_Form &$form) {
+    $this->form = $form;
+    $this->templatePath = CRM_MembershipExtras_ExtensionUtil::path() . '/templates';
+  }
+
+  /**
+   * This method should check if the given action is the proper operation so
+   * that the form is changed to have payment plan data (ie. new for membership
+   * form, renew for membership renewal form).
+   *
+   * @param int $action
+   *   Constant value for the action, as defined in CRM_Core_Action class
+   *
+   * @return mixed
+   */
+  protected abstract function isCorrectOperation($action);

```

Comment:
I think isAllowedOperation is more clear name than "Correct"

Reply:
Allowed is not the adequate concept. For example, a user may be allowed to create new memberships, but that's not what I need to verify. What I need to check is if the current action is the action that would require the change, as the same form can be used to 'view', 'update' or 'delete'. I need to make sure action is either 'new' or 'renew', respectively. 

I had to implement this, because the code was getting executed when I tried to delete  membership, for example, as the new membership form is used both to show the delete modal, and do the actual deletion. Crazy, I know, but there's CiviCRM for you :)

So I don't need to check if the action is allowed, but I do need to check if the current action is the correct/adequate/proper one to change the form structure.

Reply:
I see, I think it make sense